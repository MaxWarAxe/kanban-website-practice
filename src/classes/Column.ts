import type { Task } from "./Task.js"

export class Column extends Object {
    public id : number = 0;
    public header : string = '';
    public items : Task[] = [];

    constructor(id : number, header : string, items : Task[]) {
        super();
        this.id = id;
        this.header = header;
        this.items = items;
    }
}