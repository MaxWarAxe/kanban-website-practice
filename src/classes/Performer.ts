export class Performer extends Object{
    public id : number;
    public name : string;

    constructor(id: number, name: string){
        super();
        this.id = id;
        this.name = name;
    }
}