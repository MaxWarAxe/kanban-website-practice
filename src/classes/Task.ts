import type { Performer } from "./Performer"

export class Task extends Object{
    public id : number;
    public title : string;
    public content : string;
    public performer : Performer;

    constructor(id: number, title: string, content: string, performer: Performer){
        super();
        this.id = id;
        this.title = title;
        this.content = content;
        this.performer = performer;
    }
}