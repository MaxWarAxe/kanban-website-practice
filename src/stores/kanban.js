import { defineStore } from "pinia"
import { ref } from 'vue'
import performersJSON from '../../public/performers.json'
import tasksJSON from '../../public/tasks.json'
import columnsJSON from '../../public/columns.json'

export const useKanbanStore = defineStore('kanban', () => { 
    let columnId = ref(3)
    let taskId = ref(3)
    let performerId = ref(3)
    let columns = ref([])
    let tasks = ref([])
    let performers = ref([])
    let dragging = ref(false)

    function init(){
        // performers.value = performersJSON
        // tasks.value = tasksJSON.map((task)=>{task.performer = performers.value.find(performer => performer.id == task.performerId); return task})
        // columns.value = columnsJSON.forEach((column)=>)
        performers.value = [{id : 1, name : 'Иван Иванчевич'},
                            {id : 2, name : 'Григорий Григориевич'},
                            {id : 3, name : 'Максим Максомов'}]

        tasks.value = [{id : 1, title : 'Задача по проге', content : '1. Сделай то 2. Сделай сё 3. Сделай это', performer: performers.value[0]},
                       {id : 2, title : 'Задача по тестированию', content : 'Напиши тесты по тому, по сему', performer : performers.value[1]},
                       {id : 3, title : 'Вводим в эксплуатацию', content : 'Текст какой-то Текст какой-то Текст какой-то Текст какой-то', performer : performers.value[2]}]
        
        
        columns.value = [{id: 1, header : 'Todo', items : [tasks.value[0]]},
                        {id : 2, header : 'Tests', items : [tasks.value[1]]},
                        {id : 3, header : 'Done', items : [tasks.value[2]]}]
                           console.log(JSON.stringify(columns.value))      
    }

    function removeItemFromColumn(itemID, columnID){
    var column = columns.value.find(column => column.id == columnID)
        column.items = column.items.filter(function(item) {
            return item.id != itemID;
        });
    }

    function addItemToColumn(itemID, columnID){
        var column = columns.value.find(column => column.id == columnID)
        column.items.push(tasks.value.find(task => task.id == itemID))
    }

    function moveItem(itemID, previousColumnID, newColumnID){
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn(){
        columns.value.push({id : ++columnId.value, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID){
        var column = columns.value.find(column => column.id == columnID)
        columns.value = columns.value.filter(function(column) {
            return column.id != columnID;
        });
    }

    function addNewItem(columnID){
        var column = columns.value.find(column => column.id == columnID)
        var newTask = {id : ++taskId.value, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event,itemID,currentColumnID){
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', itemID)
        event.dataTransfer.setData('columnID',currentColumnID)
    }

    function onDropColumn(event, columnID) {
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID, columnID){
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(function(task){
            return task.id != taskID;
        })
    }

    function onRemoveDrop(event){
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID){
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name){
        performers.value.push({id : ++performerId.value, name : name})
    }

    function removePerformer(performerID){
        tasks.value.forEach(function(task){
            if(task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(function(performer){
            return performer.id != performerID;
        })
    }

    return{
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})