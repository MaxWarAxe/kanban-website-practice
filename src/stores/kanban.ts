import { defineStore } from "pinia"
import { Ref, ref } from 'vue'
import columnsJSON from '@/assets/data/columns.json' 
import type { Column } from '@/classes/Column.ts'
import type { Task } from '@/classes/Task.ts'
import type { Performer } from '@/classes/Performer.ts'

export const useKanbanStore = defineStore('kanban', () => { 
    const columnId : Ref<number> = ref(1)
    const taskId : Ref<number> = ref(1)
    const performerId : Ref<number> = ref(1)
    const columns : Ref<Column[]> = ref(new Array<Column>)
    const tasks : Ref<Task[]> = ref(new Array<Task>)
    const performers : Ref<Performer[]> = ref(new Array<Performer>)
    const dragging : Ref<boolean> = ref(false)

    function init(){   
        columns.value = columnsJSON
        for (let i = 0; i < columns.value.length; i++) {
            for (let j = 0; j < columns.value[i].items.length; j++){
                tasks.value.push(columns.value[i].items[j])
            }
        }
        for (let i = 0; i < tasks.value.length; i++) {
            if (tasks.value[i].performer == null)
                continue
            if (!performers.value.find((performer : Performer)=>performer.id == tasks.value[i].performer.id)) {
                performers.value.push(tasks.value[i].performer)
            }   
        }
        columnId.value = Math.max(...columns.value.map((column : Column) => column.id))
        taskId.value = Math.max(...tasks.value.map((task : Task) => task.id))
        performerId.value = Math.max(...performers.value.map((performer : Performer) => performer.id))
    }

    function removeItemFromColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id === columnID)
        column.items = column.items.filter((item : Task) => item.id !== itemID)
    }

    function addItemToColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        column.items.push(tasks.value.find((task : Task) => task.id == itemID))
    }

    function moveItem(itemID : number, previousColumnID : number, newColumnID : number) {
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn() {
        const newColumnID = columnId.value++
        columns.value.push({id : newColumnID, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID : number) {
        columns.value = columns.value.filter((column : Column) => column.id != columnID)
    }

    function addNewItem(columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        const newTaskId = taskId.value++
        const newTask = {id : newTaskId, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event : DragEvent, itemID : number,currentColumnID : number) {
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', String(itemID))
        event.dataTransfer.setData('columnID',String(currentColumnID))
    }

    function onDropColumn(event : DragEvent, columnID : number) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID : number, columnID : number) {
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(task => task.id != taskID)
    }

    function onRemoveDrop(event : DragEvent) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID : number) {
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name : string) {
        const newPerformerId = performerId.value++
        performers.value.push({id : newPerformerId, name : name})
    }

    function removePerformer(performerID : number) {
        tasks.value.forEach(function(task){
            if (task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(performer=>performer.id != performerID)
    }

    return {
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})