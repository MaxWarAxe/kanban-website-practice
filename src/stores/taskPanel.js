import { defineStore } from "pinia"
import { ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    let visible = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {visible,show,close}
})