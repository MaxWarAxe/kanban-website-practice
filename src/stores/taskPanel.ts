import { defineStore } from "pinia"
import { Ref, ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    const visible : Ref<boolean> = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {
        visible,
        show,
        close
    }
})