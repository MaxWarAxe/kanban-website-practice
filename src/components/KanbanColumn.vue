<script setup lang="ts">
    import { useKanbanStore } from '@/stores/kanban.ts'
    import { ModelRef } from 'vue';
    const kanbanStore = useKanbanStore()
    
    const props = defineProps<{
        id : number
    }>()

    let headerName : ModelRef<string> = defineModel()
</script>

<template>
    <div class="column-container">
        <header class="column-container__header">
            <input class="column-container__header-text column-container__header-input" 
                   placeholder="Без названия" 
                   v-model="headerName"
                   data-test="header-input"
                   />
            <svg @click="kanbanStore.addNewItem(props.id)" 
                  class="column-container__icon" 
                  viewBox="0 0 24 24" width="32" 
                  height="32"
                  data-test="add-button"
                  >
                <use href="@/assets/icons/themeIcon.svg#plus-icon"/>
            </svg>
            <svg @click="kanbanStore.removeColumn(props.id)" 
                 class="column-container__icon"
                 viewBox="0 0 24 24" 
                 width="32" 
                 height="32"
                 data-test="remove-button"
                 >
                <use href="@/assets/icons/themeIcon.svg#trash-icon"/>
            </svg>
        </header>
        <main dropzone 
              @dragover.prevent 
              @dragenter.prevent 
              @dragleave.prevent 
              class="column-container__content"
              >
            <slot data-test="task"></slot>      
        </main>
    </div>
</template>

<style lang="scss" scoped>
    $border_radius : 10px;
    $header_padding : 30px;

    .column-container{
        width: 400px;
        min-width: 400px;
        height: auto; 
        display: flex;
        flex-direction: column;
        background-color: var(--primary-container-color);
        border-radius: $border_radius;
     
        &__header{
            height: 80px;
            width: 100%;
            display: flex;
            justify-content: start;
            align-items: center;
            padding-left: $header_padding;
            padding-right: $header_padding;
            font-size: 24px;
            font-weight: bold;
            gap: 20px;
            color: var(--secondary-text-color);
            background-color: var(--primary-selection-color);
            border-radius: $border_radius $border_radius 0px 0px;
        }

        &__header-text{
            margin-right: auto;
        }

        &__header-input{
            font-size: 30px;
            width: 100%;
            color: var(--secondary-text-color);
            background-color: transparent;
            border: 0;
        }
        
        &__content{
            width: 100%;
            min-height: 260px;
            display: flex;
            max-height: 780px;
            overflow-y: auto;
            overflow-x: hidden;
            scrollbar-width: thin;
            gap: 30px;
            padding: 30px;
            align-items: center;
            flex-direction: column;
        }
        
        &__icon{
            fill: var(--secondary-text-color);
            width: 48px;
            height: 48px;
            cursor: pointer;
        }
    }
    
    
   
</style>