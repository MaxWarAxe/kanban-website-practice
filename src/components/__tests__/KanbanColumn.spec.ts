import { describe,it,expect,beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { setActivePinia, createPinia } from 'pinia'
import { vi } from 'vitest'
import KanbanColumn from '@/components/KanbanColumn.vue'
import { useKanbanStore } from '@/stores/kanban'
import { nextTick } from 'vue'
import { Column } from '@/classes/Column'
import { Task } from '@/classes/Task'
describe('KanbanColumn tests', () => {

    beforeEach(() => {
        setActivePinia(createPinia())
    })

    it('Коректно рендерится', ()=>{
        const wrapper = mount(KanbanColumn, {
            global: {
                plugins: [createTestingPinia({createSpy:vi.fn})]
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('Добавляется задача (store)',async ()=>{
        const wrapper = mount(KanbanColumn, {
            global: {
                plugins: [createTestingPinia({createSpy:vi.fn,stubActions: false})]
            },
            props : {
                id: 1,
            }
        })

        const kanbanStore = useKanbanStore()
        kanbanStore.columns.push(new Column(1,'Тестовое название', []))
        wrapper.get('[data-test="add-button"]').trigger('click')
        await nextTick()
        expect(kanbanStore.tasks.length).toBe(1)
    })

    it('Удаление колонки (store)',async ()=>{
        const wrapper = mount(KanbanColumn, {
            global: {
                plugins: [createTestingPinia({createSpy:vi.fn,stubActions: false})]
            },
            props : {
                id: 1,
            }
        })

        const kanbanStore = useKanbanStore()
        kanbanStore.columns.push(new Column(1,'Тестовое название', []))
        wrapper.get('[data-test="add-button"]').trigger('click')
        await nextTick()
        const removeButton = wrapper.find('[data-test="remove-button"]')
        removeButton.trigger('click')
        console.log(wrapper.html())
        expect(kanbanStore.columns.length).toBe(0)
    })
})