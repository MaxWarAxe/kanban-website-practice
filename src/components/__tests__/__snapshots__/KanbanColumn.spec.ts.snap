// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KanbanColumn tests > Коректно рендерится 1`] = `
"<div data-v-5fec18df="" class="column-container">
  <header data-v-5fec18df="" class="column-container__header"><input data-v-5fec18df="" class="column-container__header-text column-container__header-input" placeholder="Без названия" data-test="header-input"><svg data-v-5fec18df="" class="column-container__icon" viewBox="0 0 24 24" width="32" height="32" data-test="add-button">
      <use data-v-5fec18df="" href="/src/assets/icons/themeIcon.svg#plus-icon"></use>
    </svg><svg data-v-5fec18df="" class="column-container__icon" viewBox="0 0 24 24" width="32" height="32" data-test="remove-button">
      <use data-v-5fec18df="" href="/src/assets/icons/themeIcon.svg#trash-icon"></use>
    </svg></header>
  <main data-v-5fec18df="" dropzone="" class="column-container__content"></main>
</div>"
`;
